ios/keyman/Keyman/Keyman/MainViewController.swift

```diff
@@ -0,0 +1,1200 @@
+    let view =  barButtonItem.value(forKey: "view") as? UIView
+    return view?.frame ?? CGRect.zero
+  }
+
+  func startTimer() {
+    if loadTimer == nil {
+      loadTimer = Timer.scheduledTimer(timeInterval: 1.0, target: self,
+                                       selector: #selector(self.timerAction), userInfo: nil, repeats: true)
+    }
+  }
+
+  func stopTimer() {
+    if let timer = loadTimer {
+      timer.invalidate()
+      loadTimer = nil
+    }
+  }
+
+  func timerAction() {
+    if updateStatus == 1 {
+      stopTimer()
+
+      dismissActivityIndicator()
+
+      textView.isEditable = true
+      textView.becomeFirstResponder()
+
+      if let launchUrl = launchUrl {
+        performAction(from: launchUrl)
+      } else {
+        loadSavedUserText()
+        if shouldShowGetStarted() {
+          perform(#selector(self.showGetStartedView), with: nil, afterDelay: 1.0)
+        }
+      }
+    } else if updateStatus == -1 {
+      stopTimer()
+      dismissActivityIndicator()
+      loadSavedUserText()
+      textView.isEditable = true
+      textView.becomeFirstResponder()
+      if shouldShowGetStarted() {
+        perform(#selector(self.showGetStartedView), with: nil, afterDelay: 1.0)
+      }
+    }
+  }
+
+  func resetTextViewCursor() {
+    textView.selectedRange = NSRange(location: 0, length: 0)
+
+    // TODO: Figure out what this does
+    if let size = textView.font?.pointSize {
+      textView.font = textView.font?.withSize(size + 5)
+      textView.font = textView.font?.withSize(size - 5)
+    }
+  }
+
+  func loadSavedUserText() {
+    let userData = AppDelegate.activeUserDefaults()
+    let userText = userData.object(forKey: userTextKey) as? String
+    let userTextSize = userData.object(forKey: userTextSizeKey) as? String
+
+    if let text = userText, !text.isEmpty {
+      textView.text = text
+    }
+
+    if let sizeString = userTextSize, let size = Float(sizeString) {
+      textSize = CGFloat(size)
+      textView.font = textView.font?.withSize(textSize)
+      textSizeController.value = (Float(textSize - minTextSize))
+    }
+  }
+
+  func params(of query: String) -> [String: String] {
+    let components = query.components(separatedBy: "&")
+    return components.reduce([String: String]()) { prevParams, s in
+      var params = prevParams
+      let components = s.components(separatedBy: "=")
+      if components.count == 2 {
+        params[components[0]] = components[1]
+      }
+      return params
+    }
+  }
+
+  // TODO: Refactor control flow
+  func performAction(from url: URL) {
+    guard let query = url.query else {
+      launchUrl = nil
+      return
+    }
+
+    if url.lastPathComponent != "open" {
+      return
+    }
+
+    let params = self.params(of: query)
+    if params["url"] == nil {
+      // Download and set custom keyboard
+
+      let jsonUrl = params["url"].flatMap { URL(string: $0) }
+      let direct = params["direct"]
+
```

Comment:
There were two other values retrieved here, but it seems they were unused in this branch of the control flow.